packages/frontend/editor-ui/src/features/dataStore/composables/useDataStoreOperations.ts

```diff
@@ -0,0 +1,295 @@
+	dataStoreId: string;
+	projectId: string;
+	gridApi: Ref<GridApi>;
+	totalItems: Ref<number>;
+	setTotalItems: (count: number) => void;
+	ensureItemOnPage: (itemIndex: number) => Promise<void>;
+	focusFirstEditableCell: (rowId: number) => void;
+	toggleSave: (value: boolean) => void;
+	currentPage: Ref<number>;
+	pageSize: Ref<number>;
+	currentSortBy: Ref<string>;
+	currentSortOrder: Ref<string | null>;
+	handleClearSelection: () => void;
+	selectedRowIds: Ref<Set<number>>;
+	handleCopyFocusedCell: (params: CellKeyDownEvent<DataStoreRow>) => Promise<void>;
+};
+
+export const useDataStoreOperations = ({
+	colDefs,
+	rowData,
+	deleteGridColumn,
+	addGridColumn,
+	setGridData,
+	insertGridColumn,
+	moveGridColumn,
+	dataStoreId,
+	projectId,
+	gridApi,
+	totalItems,
+	setTotalItems,
+	ensureItemOnPage,
+	focusFirstEditableCell,
+	toggleSave,
+	currentPage,
+	pageSize,
+	currentSortBy,
+	currentSortOrder,
+	handleClearSelection,
+	selectedRowIds,
+	handleCopyFocusedCell,
+}: UseDataStoreOperationsParams) => {
+	const i18n = useI18n();
+	const toast = useToast();
+	const message = useMessage();
+	const dataStoreStore = useDataStoreStore();
+	const contentLoading = ref(false);
+
+	async function onDeleteColumn(columnId: string) {
+		const columnToDelete = colDefs.value.find((col) => col.colId === columnId);
+		if (!columnToDelete) return;
+
+		const promptResponse = await message.confirm(
+			i18n.baseText('dataStore.deleteColumn.confirm.message', {
+				interpolate: { name: columnToDelete.headerName ?? '' },
+			}),
+			i18n.baseText('dataStore.deleteColumn.confirm.title'),
+			{
+				confirmButtonText: i18n.baseText('generic.delete'),
+				cancelButtonText: i18n.baseText('generic.cancel'),
+			},
+		);
+
+		if (promptResponse !== MODAL_CONFIRM) {
+			return;
+		}
+
+		const columnToDeleteIndex = colDefs.value.findIndex((col) => col.colId === columnId);
+		deleteGridColumn(columnId);
+		const rowDataOldValue = [...rowData.value];
+		rowData.value = rowData.value.map((row) => {
+			const { [columnToDelete.field!]: _, ...rest } = row;
+			return rest;
+		});
+		setGridData({ rowData: rowData.value });
+		try {
+			await dataStoreStore.deleteDataStoreColumn(dataStoreId, projectId, columnId);
+		} catch (error) {
+			toast.showError(error, i18n.baseText('dataStore.deleteColumn.error'));
+			insertGridColumn(columnToDelete, columnToDeleteIndex);
+			rowData.value = rowDataOldValue;
+			setGridData({ rowData: rowData.value });
+		}
+	}
+
+	async function onAddColumn(column: DataStoreColumnCreatePayload) {
+		try {
+			const newColumn = await dataStoreStore.addDataStoreColumn(dataStoreId, projectId, column);
+			addGridColumn(newColumn);
+			rowData.value = rowData.value.map((row) => {
+				return { ...row, [newColumn.name]: null };
+			});
+			setGridData({ rowData: rowData.value });
+			return true;
+		} catch (error) {
+			toast.showError(error, i18n.baseText('dataStore.addColumn.error'));
```

Comment:
If this fails, should we remove column from UI?

Reply:
The thing that we're trying to handle failure of is the api call which is done first so in practice the column shouldn't be added to the UI in that case. Having just the api call in the try block in would be more correct but then we'll have to deal with the ts complaints..